### PR TITLE
ScsSolver defaults not to print

### PR DESCRIPTION
**Pull request by hongkai-dai**
cc @RussTedrake 
As mentioned in https://github.com/RobotLocomotion/drake/pull/15251#issuecomment-870105746

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@jwnimmer-tri for feature review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257#-:-MdLDqfA49dDa0zpgUzb:b-o4a2ka)* status: LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri)

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm:

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257#-:-MdLGgXx2ZNCH5CsOgM3:bnfp4nl)* status: 2 unresolved discussions, needs at least two assigned reviewers (waiting on @hongkai-dai)

___
*[solvers/scs_solver.cc, line 613 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdLEWC-DcqF4agfOVfF:-MdLEWC0-fDGEvN3hO1q:bd51x0s) ([raw file](https://github.com/robotlocomotion/drake/blob/9b9a114bef466beff14f987a03e5e62e3cf95932/solvers/scs_solver.cc#L613)):*
> ```cpp
> }
> 
> void SetScsSettings(
> ```

nit The GSG says that we should only use overloading when the function is doing the same thing in all cases.

https://drake.mit.edu/styleguide/cppguide.html#Function_Overloading

This function seems different enough to me to warrant a new name.  We're not setting SCS's native options structure directly, rather instead we're re-keying a common options dict into a temporary options dict.

(Writing API documentation explaining what this function does would also help.)
___
*[solvers/test/scs_solver_test.cc, line 240 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdLG42FBR3XtI-d3-Ym:-MdLG42FBR3XtI-d3-Yn:b-79usrw) ([raw file](https://github.com/robotlocomotion/drake/blob/9b9a114bef466beff14f987a03e5e62e3cf95932/solvers/test/scs_solver_test.cc#L240)):*
> ```cpp
>   ScsSolver solver;
>   if (solver.available()) {
>     prob()->prog()->SetSolverOption(ScsSolver::id(), "verbose", 0);
> ```

nit Is this line useful anymore?

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: needs at least two assigned reviewers (waiting on @jwnimmer-tri)

___
*[solvers/test/scs_solver_test.cc, line 240 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdLG42FBR3XtI-d3-Ym:-MdNVpO62TpUnIS8Y-04:b-376vyo) ([raw file](https://github.com/robotlocomotion/drake/blob/9b9a114bef466beff14f987a03e5e62e3cf95932/solvers/test/scs_solver_test.cc#L240)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit Is this line useful anymore?
</blockquote></details>

Done. Good catch!

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: feature.

+@SeanCurtis-TRI for platform review per schedule, please.

> Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257#-:-MdNW0BOAU6nnJT9FlDP:b-tdex0x)* status: 1 unresolved discussion, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai and @SeanCurtis-TRI)

___
*[solvers/test/scs_solver_test.cc, line 403 at r3](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNW7p0BIBgMbyepu_D:-MdNW7p139k_e4W74FHl:biwbs78) ([raw file](https://github.com/robotlocomotion/drake/blob/d980ba256469932ab374489b390d7a506dc17ff6/solvers/test/scs_solver_test.cc#L403)):*
> ```cpp
> }
> 
> GTEST_TEST(TestScs, TestVerbose) {
> ```

nit In the other PR, we decided to clarify this test purpose.  Maybe add it here, also?

```c++
// This is a code coverage test, not a functional test.  If the code that
// handles verbosity options has a segfault or always throws an exception,
// then this would catch it.
```

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @SeanCurtis-TRI)

**Review by SeanCurtis-TRI — COMMENTED** [review]
I have a question about the semantics of `kPrintToConsole` and its impact on solvers. How uniform is its semantics supposed to be? It is called "common", but it works differently in the scs solver than it does for the clp. I want to make sure that's clear and desired (and, at least, documented).

> Reviewed 1 of 1 files at r2, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257#-:-MdNmdc-0HQBPHenmNLy:b-reftd3)* status: 2 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai)

___
*[solvers/scs_solver.cc, line 619 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNj2Vb9Gz5axjMGXbJ:-MdNj2Vb9Gz5axjMGXbK:bifs4at) ([raw file](https://github.com/robotlocomotion/drake/blob/6c357a0e3b68147aa909ea7b30b54b7b107ad835/solvers/scs_solver.cc#L619)):*
> ```cpp
>         common_options,
>     std::unordered_map<std::string, int>* input_solver_options_int,
>     std::unordered_map<std::string, double>* input_solver_options_double) {
> ```

BTW Rather than

```c++
...
std::unordered_map<std::string, double>* input_solver_options_double) {
  unused(input_solver_options_double);
...
}
```

You could simply omit the name of the parameter:

```c++
...
std::unordered_map<std::string, double>*) {
...
}
```
___
*[solvers/scs_solver.cc, line 621 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNjosP-SW5COfF23sB:-MdNjosP-SW5COfF23sC:ble0spq) ([raw file](https://github.com/robotlocomotion/drake/blob/6c357a0e3b68147aa909ea7b30b54b7b107ad835/solvers/scs_solver.cc#L621)):*
> ```cpp
>     std::unordered_map<std::string, double>* input_solver_options_double) {
>   unused(input_solver_options_double);
>   auto it = common_options.find(CommonSolverOption::kPrintToConsole);
> ```

BTW I found this organization *slightly* confusing.

This method *sometimes* sets the solver settings ("verbose") depending on whether `kPrintToConsole` has been set in the `common_options`. This method doesn't suggest any default behavior.

In the invocation of `SetScsSettings` that consumes the solver settings possibly set here, we have the default logic (if "verbose" *hasn't* been specified at all, then we default to off.

As I understand it, the logic vis a vis `kPrintToConsole` should be:

 1. If solver setting "verbose" has already been set, `kPrintToConsole` is ignored.
 2. No `kPrintToConsole` specification implies -> no output
 3. `kPrintToConsole` specified to off --> no output
 4. `kPrintToConsole` specified to on --> output

If that's the case, it's easier to reason about/validate that logic if it all happens in a single method.

Then all the logic is locally encapsulated and this also facilitates testing the Drake logic (if we had friend access to this method specifically).
___
*[solvers/scs_solver.cc, line 626 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNlxCK80ndGbaAepW7:-MdNlxCK80ndGbaAepW8:bgkb917) ([raw file](https://github.com/robotlocomotion/drake/blob/6c357a0e3b68147aa909ea7b30b54b7b107ad835/solvers/scs_solver.cc#L626)):*
> ```cpp
>     DRAKE_DEMAND(option_value == 0 || option_value == 1);
>     auto it_verbose = input_solver_options_int->find("verbose");
>     if (it_verbose == input_solver_options_int->end()) {
> ```

BTW It seems this could lead to some confusion. If the solver has "verbose" set on in one place in code, but the user sets the common `kPrintToConsole` to off elsewhere, it will have no impact. I can imagine that would confuse the end user. Is this behavior documented?

Would it be appropriate to provide a `static const drake::logging::Warn log_once("")` arrangement in the case where they have explicitly set both?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 2 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai)

___
*[solvers/scs_solver.cc, line 626 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNlxCK80ndGbaAepW7:-MdO3KHqAwNQvjp0B3Q2:b-6r6rlc) ([raw file](https://github.com/robotlocomotion/drake/blob/6c357a0e3b68147aa909ea7b30b54b7b107ad835/solvers/scs_solver.cc#L626)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW It seems this could lead to some confusion. If the solver has "verbose" set on in one place in code, but the user sets the common `kPrintToConsole` to off elsewhere, it will have no impact. I can imagine that would confuse the end user. Is this behavior documented?

Would it be appropriate to provide a `static const drake::logging::Warn log_once("")` arrangement in the case where they have explicitly set both?
</blockquote></details>

I don't think we should add runtime warning, but clarifying the documentation would be good.  It's a folder-wide policy, not anything SCS specific, so probably on SolverOptions and/or CommonSolverOptions class overview.

The priority given to solver options is as follows (from lowest / least, to highest / most):

1. common option set on the MathematicalProgram itself
2. common option passed as an argument to Solve
3. solver-specific option set on the MathematicalProgram itself
4. solver-specific option passed as an argument to Solve

The idea is that the more specific and local the option is, the higher priority it gets.  You could imagine setting kPrint=1 on the MathematicalProgram by default in some unit test (tier 1), but there might be really spammy solve that we want to disable in certain cases (tier 4) in case it gets chose.  In that case, it is as-desired that the verbosity would be off, even though kPrint was toggled at tier 1.

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 2 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai)

___
*[solvers/scs_solver.cc, line 626 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNlxCK80ndGbaAepW7:-MdOBm1Y16Yrzwy-2cNm:b-3gp1v2) ([raw file](https://github.com/robotlocomotion/drake/blob/6c357a0e3b68147aa909ea7b30b54b7b107ad835/solvers/scs_solver.cc#L626)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't think we should add runtime warning, but clarifying the documentation would be good.  It's a folder-wide policy, not anything SCS specific, so probably on SolverOptions and/or CommonSolverOptions class overview.

The priority given to solver options is as follows (from lowest / least, to highest / most):

1. common option set on the MathematicalProgram itself
2. common option passed as an argument to Solve
3. solver-specific option set on the MathematicalProgram itself
4. solver-specific option passed as an argument to Solve

The idea is that the more specific and local the option is, the higher priority it gets.  You could imagine setting kPrint=1 on the MathematicalProgram by default in some unit test (tier 1), but there might be really spammy solve that we want to disable in certain cases (tier 4) in case it gets chose.  In that case, it is as-desired that the verbosity would be off, even though kPrint was toggled at tier 1.
</blockquote></details>

I'd certainly consider clear documentation a suitable way to resolve this.

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai)

___
*[solvers/scs_solver.cc, line 621 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNjosP-SW5COfF23sB:-MdPC8V37AHvIlK_DmsG:b-ktjcxh) ([raw file](https://github.com/robotlocomotion/drake/blob/6c357a0e3b68147aa909ea7b30b54b7b107ad835/solvers/scs_solver.cc#L621)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I found this organization *slightly* confusing.

This method *sometimes* sets the solver settings ("verbose") depending on whether `kPrintToConsole` has been set in the `common_options`. This method doesn't suggest any default behavior.

In the invocation of `SetScsSettings` that consumes the solver settings possibly set here, we have the default logic (if "verbose" *hasn't* been specified at all, then we default to off.

As I understand it, the logic vis a vis `kPrintToConsole` should be:

 1. If solver setting "verbose" has already been set, `kPrintToConsole` is ignored.
 2. No `kPrintToConsole` specification implies -> no output
 3. `kPrintToConsole` specified to off --> no output
 4. `kPrintToConsole` specified to on --> output

If that's the case, it's easier to reason about/validate that logic if it all happens in a single method.

Then all the logic is locally encapsulated and this also facilitates testing the Drake logic (if we had friend access to this method specifically).

</blockquote></details>

Done. I removed one function, now the logic for print out option is just in SetScsSettings.
___
*[solvers/scs_solver.cc, line 626 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNlxCK80ndGbaAepW7:-MdOY55X86ZAfFKFBuRd:bapf30m) ([raw file](https://github.com/robotlocomotion/drake/blob/6c357a0e3b68147aa909ea7b30b54b7b107ad835/solvers/scs_solver.cc#L626)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'd certainly consider clear documentation a suitable way to resolve this.
</blockquote></details>

Good call. I added the documentation in `solve.h`

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 1 unresolved discussion, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai)

___
*[solvers/scs_solver.cc, line 573 at r5](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdPD2cB4eTdMqXaFNZB:-MdPD2cB4eTdMqXaFNZC:bv7zx0d) ([raw file](https://github.com/robotlocomotion/drake/blob/3ee3d4f53f787f38514d5d8b0d31b939b11e0792/solvers/scs_solver.cc#L573)):*
> ```cpp
>       scs_settings->verbose = print_to_console_val;
>     } else {
>     // Default to no print.
> ```

nit Needs +2 indentation for else body.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai)

**Review by SeanCurtis-TRI — COMMENTED** [review]
More questions. :/

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257#-:-MdSTajP7BKTn1f5sB3a:b-n4q5u)* status: 2 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai)

___
*[solvers/scs_solver.cc, line 626 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNlxCK80ndGbaAepW7:-MdSPn1REB09t26LxZwz:b8hfr1t) ([raw file](https://github.com/robotlocomotion/drake/blob/6c357a0e3b68147aa909ea7b30b54b7b107ad835/solvers/scs_solver.cc#L626)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Good call. I added the documentation in `solve.h`
</blockquote></details>

I don't see any changes to `solve.h` in this PR.
___
*[solvers/scs_solver.cc, line 545 at r6](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdSQpU_DHLjnNAqsiDH:-MdSQpU_DHLjnNAqsiDI:bb1kij0) ([raw file](https://github.com/robotlocomotion/drake/blob/1878a3c5617ed0a10b59ecfee575b4d111d48040/solvers/scs_solver.cc#L545)):*
> ```cpp
> 
> namespace {
> void SetScsSettings(
> ```

As a person largely unfamiliar with this corner of Drake, something jumped out at me when reading this logic that wasn't immediately apparent.

Upon reading this method, I *imagined* a scenario in which this method got invoked twice: the first time with "verbose" defined in solver settings and the second time with `kPrintToConsole` defined (but no solver settings). In that case, the earlier invocation would be undone and the priority protocol would be subverted.

Upon deeper analysis, I recognized that this method is only called in the `DoSolve` method which takes a complete set of merged settings. And the duration for which those settings apply is simply the duration of that invocation of `DoSolve`. So, the scenario I described above isn't actually possible.

The correctness of this method's ability to implement the desired priority protocol is predicated on when and how it is called. It seems that should be documented to alleviate confusion in readers in the future.
___
*[solvers/scs_solver.cc, line 562 at r6](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdSSZURBm2kWZ56YV7f:-MdSSZURBm2kWZ56YV7g:b-z6avoo) ([raw file](https://github.com/robotlocomotion/drake/blob/1878a3c5617ed0a10b59ecfee575b4d111d48040/solvers/scs_solver.cc#L562)):*
> ```cpp
>   }
>   it = solver_options_int->find("verbose");
>   if (it != solver_options_int->end()) {
> ```

Use case question:

Imagine I have a mathematical program and solver. Both are configured to have verbosity turned off.

The first time I invoke the solver, I pass in settings that set "verbose" to 1 (doesn't matter how). The second time I invoke the solver, I pass in a settings that contain *neither* solver settings for "verbose" nor common settings for `kPrintToConsole`.

What is the expected result? Am I printing to the console or not? Do my settings *only* apply to the duration of `DoSolve`? Or do they definitely persist and the only way I can go back to "default" behavior is to set it myself?

In other words, in terms of implementation, should this method configure things to their default configuration on every invocation and only overwrite that if the settings explicitly say something?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 2 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai and @SeanCurtis-TRI)

___
*[solvers/scs_solver.cc, line 562 at r6](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdSSZURBm2kWZ56YV7f:-MdSdu31CKo4OyaqqT3A:b-6jp4ye) ([raw file](https://github.com/robotlocomotion/drake/blob/1878a3c5617ed0a10b59ecfee575b4d111d48040/solvers/scs_solver.cc#L562)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Use case question:

Imagine I have a mathematical program and solver. Both are configured to have verbosity turned off.

The first time I invoke the solver, I pass in settings that set "verbose" to 1 (doesn't matter how). The second time I invoke the solver, I pass in a settings that contain *neither* solver settings for "verbose" nor common settings for `kPrintToConsole`.

What is the expected result? Am I printing to the console or not? Do my settings *only* apply to the duration of `DoSolve`? Or do they definitely persist and the only way I can go back to "default" behavior is to set it myself?

In other words, in terms of implementation, should this method configure things to their default configuration on every invocation and only overwrite that if the settings explicitly say something?
</blockquote></details>

To the question in your last paragraph -- yes, that's correct; solvers have no memory.  You can feed them a new program to solve each time, and the configuration of the prior solved program does not bleed over into the new program.

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 2 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai and @jwnimmer-tri)

___
*[solvers/scs_solver.cc, line 562 at r6](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdSSZURBm2kWZ56YV7f:-MdSf14z33_UT8ooJIM1:bard5wx) ([raw file](https://github.com/robotlocomotion/drake/blob/1878a3c5617ed0a10b59ecfee575b4d111d48040/solvers/scs_solver.cc#L562)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

To the question in your last paragraph -- yes, that's correct; solvers have no memory.  You can feed them a new program to solve each time, and the configuration of the prior solved program does not bleed over into the new program.
</blockquote></details>

That was my hope/expectation. In that case, I don't believe that this code does that.

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 2 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai and @SeanCurtis-TRI)

___
*[solvers/scs_solver.cc, line 562 at r6](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdSSZURBm2kWZ56YV7f:-MdSaN3t5wR0K6oFwenB:bd01ik5) ([raw file](https://github.com/robotlocomotion/drake/blob/1878a3c5617ed0a10b59ecfee575b4d111d48040/solvers/scs_solver.cc#L562)):*
> Do my settings only apply to the duration of DoSolve? 

It depends on how the options are set. There are 4 ways to set a print option
1. Common option stored in MathematicalProgram, via 
    ```
    SolverOptions solver_options;
    solver_options.SetOption(CommonSolverOption::kPrintToConsole, 1);
    prog.SetSolverOptions(solver_options);
    Solve(prog);
    ```
    the solver options is stored inside `prog`. Next time you call Solve, unless you change the options via approach 2, 3, 4, the solver will print to console.
2. Common option passed as an argument to `Solve`
    ```
    SolverOptions solver_options;
    solver_options.SetOption(CommonSolverOption::kPrintToConsole, 1);
    Solve(prog, std::nullopt, solver_options);
    ```
    the solver options is not stored inside `prog`. Next time you call Solve, it will use whatever option you set for your next call.
3. solver-specific option set on the MathematicalProgram itself
    ```
    prog.SetSolverOption(ScsSolver::id(), "verbose", 1);
    Solve(prog);
    ```
    the solver option is stored inside `prog`. Next time you call `Solve`, unless you change the options via approach 4, the solver will print to console.
4. solver-specific option passed as an argument to Solve
    ```
    SolverOptions solver_options;
    solver_options.SetOption(ScsSolver::id(), "verbose", 1);
    Solver(prog, std::nullopt, solver_options);
   ```
    the solver option is not stored inside `prog`. The next time you call `Solve`, it will use whatever solver options you pass in the next call.
    ```
```

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 1 unresolved discussion, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @hongkai-dai)

___
*[solvers/scs_solver.cc, line 562 at r6](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdSSZURBm2kWZ56YV7f:-MdSgMTZ3_G80VH8osbq:b4i4q5u) ([raw file](https://github.com/robotlocomotion/drake/blob/1878a3c5617ed0a10b59ecfee575b4d111d48040/solvers/scs_solver.cc#L562)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

> Do my settings only apply to the duration of DoSolve? 

It depends on how the options are set. There are 4 ways to set a print option
1. Common option stored in MathematicalProgram, via 
    ```
    SolverOptions solver_options;
    solver_options.SetOption(CommonSolverOption::kPrintToConsole, 1);
    prog.SetSolverOptions(solver_options);
    Solve(prog);
    ```
    the solver options is stored inside `prog`. Next time you call Solve, unless you change the options via approach 2, 3, 4, the solver will print to console.
2. Common option passed as an argument to `Solve`
    ```
    SolverOptions solver_options;
    solver_options.SetOption(CommonSolverOption::kPrintToConsole, 1);
    Solve(prog, std::nullopt, solver_options);
    ```
    the solver options is not stored inside `prog`. Next time you call Solve, it will use whatever option you set for your next call.
3. solver-specific option set on the MathematicalProgram itself
    ```
    prog.SetSolverOption(ScsSolver::id(), "verbose", 1);
    Solve(prog);
    ```
    the solver option is stored inside `prog`. Next time you call `Solve`, unless you change the options via approach 4, the solver will print to console.
4. solver-specific option passed as an argument to Solve
    ```
    SolverOptions solver_options;
    solver_options.SetOption(ScsSolver::id(), "verbose", 1);
    Solver(prog, std::nullopt, solver_options);
   ```
    the solver option is not stored inside `prog`. The next time you call `Solve`, it will use whatever solver options you pass in the next call.
    ```
</blockquote></details>

Oh...the key is that everything is getting written into a `SRC_SETTINGS`. And that is being instantiated anew in each call to `DoSolve`. Sorry, my mental model had it writing more directly to the solver. I'm good.


<!-- Sent from Reviewable.io -->

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 1 unresolved discussion, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @SeanCurtis-TRI)

___
*[solvers/scs_solver.cc, line 545 at r6](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdSQpU_DHLjnNAqsiDH:-MdSgjjyEYzdNkGRMSJ_:bc88kz7) ([raw file](https://github.com/robotlocomotion/drake/blob/1878a3c5617ed0a10b59ecfee575b4d111d48040/solvers/scs_solver.cc#L545)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

As a person largely unfamiliar with this corner of Drake, something jumped out at me when reading this logic that wasn't immediately apparent.

Upon reading this method, I *imagined* a scenario in which this method got invoked twice: the first time with "verbose" defined in solver settings and the second time with `kPrintToConsole` defined (but no solver settings). In that case, the earlier invocation would be undone and the priority protocol would be subverted.

Upon deeper analysis, I recognized that this method is only called in the `DoSolve` method which takes a complete set of merged settings. And the duration for which those settings apply is simply the duration of that invocation of `DoSolve`. So, the scenario I described above isn't actually possible.

The correctness of this method's ability to implement the desired priority protocol is predicated on when and how it is called. It seems that should be documented to alleviate confusion in readers in the future.
</blockquote></details>

Done. I added the documentation.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 1 unresolved discussion, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @SeanCurtis-TRI)

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: 1 unresolved discussion, LGTM missing from assignee SeanCurtis-TRI(platform) (waiting on @SeanCurtis-TRI)

___
*[solvers/scs_solver.cc, line 626 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdNlxCK80ndGbaAepW7:-MdSh4ZhACjlfkXq6wji:bjunw4u) ([raw file](https://github.com/robotlocomotion/drake/blob/6c357a0e3b68147aa909ea7b30b54b7b107ad835/solvers/scs_solver.cc#L626)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I don't see any changes to `solve.h` in this PR.
</blockquote></details>

Done for real

**Review by SeanCurtis-TRI — COMMENTED** [review]
:LGTM:

> Reviewed 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257#-:-MdSkLLE88LL3iyAG_rx:bmxilld)* status: :shipit: complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),SeanCurtis-TRI(platform) (waiting on @hongkai-dai)

___
*[solvers/scs_solver.cc, line 545 at r6](https://reviewable.io/reviews/robotlocomotion/drake/15257#-MdSQpU_DHLjnNAqsiDH:-MdSkEaHABW3E85daCjk:b-c2jya6) ([raw file](https://github.com/robotlocomotion/drake/blob/1878a3c5617ed0a10b59ecfee575b4d111d48040/solvers/scs_solver.cc#L545)):*
BTW Alternatively, (now that I have a greater understanding):

>  This should be invoked only once on each unique instance of `SCS_SETTINGS`.

(Here and below...but what you have is sufficient.)

**Review by SeanCurtis-TRI — COMMENTED** [review]
> Reviewed 1 of 2 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15257)* status: :shipit: complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),SeanCurtis-TRI(platform) (waiting on @hongkai-dai)
